lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -649,37 +670,50 @@ def do_3d_projection(self, renderer):
             self.update_scalarmappable()
             self._facecolors3d = self._facecolors
 
-        txs, tys, tzs = proj3d._proj_transform_vec(self._vec, renderer.M)
-        xyzlist = [(txs[sl], tys[sl], tzs[sl]) for sl in self._segslices]
+        needs_masking = self._invalid_vertices is not False
+        num_faces = len(self._faces)
+
+        pfaces = proj3d._proj_transform_vectors(self._faces, renderer.M)
+        pzs = pfaces[..., 2]
+        if needs_masking:
+            pzs = np.ma.MaskedArray(pzs, mask=self._invalid_vertices)
```

Comment:
Tempting to move this down to line 692, where it's used for the first time.

Reply:
I felt like it's logically connected to `pfaces` and it's used both on line 692, 725 and 729, so I didn't want to make the first use special in any way.

Reply:
I suggested this because it might be worth adding an optimization today along the lines of:

```
if numpy_version >= LooseVersion(1.17.0):
   # use the where argument, skip the masked array entirely
else:
   ...
```

Happy to have that declared as out of scope

Reply:
I'm happy to do that, but I'm still a bit afraid that it will make the code even less readable. I'd have to insert the same `if` on lines 692 and 729. WDYT?

Reply:
Maybe try it and see if it actually helps with performance locally in any significant way? If not, then I think it's not worth it.

Reply:
I've tried this, and the `where=` reductions are indeed faster!
```python
In [1]: import numpy as np                                            
                                                  
In [2]: maskf = np.zeros((399 * 399, 4), dtype=bool)                  
                                                  
In [3]: maskt = np.ones((399 * 399, 4), dtype=bool)                   
                         
In [4]: x = np.ones((399 * 399, 4), dtype=float)                      
                         
In [5]: mx = np.ma.MaskedArray(x, maskf)          
                                                  
In [6]: %timeit np.max(mx, axis=-1)                                   
10.3 ms ± 441 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: %timeit np.max(x, axis=-1, where=maskt, initial=0)            
7.09 ms ± 139 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

On the other hand, it looks like `np.average` doesn't support the `where=` argument, so I don't think we can do this...

Reply:
Oh, that's a good catch. You can _almost_ get away with passing the mask into the `weights` argument of np.average, but that behaves poorly on NaNs.

Reply:
Hah, good point! And it looks like it still beats the masked array implementation, which is very surprising to me:
```python
In [9]: %timeit np.average(x, axis=-1, weights=maskt)                                                
8.69 ms ± 137 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: %timeit np.average(mx, axis=-1)          
12.9 ms ± 370 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

But yes, there's an issue with invalid floating point values and there's no default argument we can pass in. I guess the latter problem could be fixed by using a wrapper like `lambda *args, **kwargs: np.average(*args, weight=kwargs.pop('where', 1.), **kwargs)` in the dict defining zsort functions.

So what's the final decision?